mutation_reader.cc

```diff
@@ -1044,9 +1044,12 @@ void shard_reader::remote_reader::adjust_partition_slice() {
         _slice_override = _ps;
     }
 
-    auto& last_ckey = _last_position_in_partition->key();
+    const auto pos = _last_position_in_partition->key().is_full(*_schema) ?
```

Comment:
I think that if you changed the semantics of `_last_position_in_partition` to mean the first position yet to be produced, you wouldn't need this transformation, and wouldn't need as much as 3 checks in `ensure_safe_progress()`. What prevents us from doing that?

Reply:
Indeed the code would be much cleaner. The problem is that the only time we actually get to peek at what is the next fragment is when the progress assurance logic kicks in, that is when the last fragment in the buffer is a `range_tombstone`. When this is not the case, `ensure_safe_progress()` returns early and we don't know what the next position is. To be able to peek the next fragment, we'd have to call `fill_buffer()` on the underlying reader, then leave that buffer there, possibly wasting the effort if the reader is evicted in the meanwhile.

If we consider reader eviction to be very reader (as it should be) then we can maybe justify always reading ahead one buffer worth of fragments, just to be able to have `_next_position_in_partition` instead of a `_last_position_in_partition`. However this would interact poorly with paging, as in the "worst" case we'd have a full buffer worth of data on each shard when the combining reader would be stopped. This is somewhat mitigated by the fact that we save readers between pages now.

Reply:
I think we can always calculate `_next_position_in_partition`. 

If the buffer is empty, don't change anything.

When the last fragment is not a range tombstone:
  - if it's a clustering row, use `after_row()`
  - if it's a partition start, use `for_static_row()`
  - if it's a static row, use `before_all_clustered_rows()`

If the last fragment is a range tombstone, set `_next_position_in_partition` to the first peeked position != last position in the buffer.

Reply:
Well, since we're only concerned with clustering positions, the partition start and static row are handled by filtering and `_next_position_in_partition` is left unchanged.